### PR TITLE
Multitouch touchscreen support

### PR DESCRIPTION
**Pull request by GregDMeyer**
These changes add support for intuitive touchscreen interaction with Matplotlib, specifically touch-drag to pan and two-finger pinch to zoom. 

When dragging, the plot is panned so that the same point in data coords stays under one's finger. When zooming, the plot is  zoomed and panned simultaneously such that the same two points in data coords stay under the two touch points (except in cases such as holding down the 'x' key, locking motion to the x axis). These behaviors are meant to mimic what one would imagine physically sliding and/or stretching something.

Touch support is similarly added for the Axes3D class, such that a drag rotates the axes, and two-finger pinch scales the plot (zooms).

I only set up the Qt4 and Qt5 backends to send touch events to the figures, mostly because I don't know enough about gtk, javascript, etc to do it for the other backends. But it shouldn't be that hard for people who know those things, because `NavigationToolbar2` does the heavy lifting. All you have to do is send it the touches' coordinates and a unique ID for each touch.

Summary of changes:

 - `NavigationToolbar2` in `backend_bases.py` has new functions for handling touches, implementing touch-drag to pan and two-finger pinch to zoom. 
 - New classes `Touch` and `TouchEvent` to store touch event data.
 - `FigureCanvasBase` has new functions to trigger the callbacks for touch events.
 - `matplotlibrc.template` and the other rc files are set up for the flag `backends.touch`, which defines whether to accept touch input. If it is false, Qt turns touches into regular mouse events.
 - Qt4 and Qt5 backends are set up to accept touch events and call the `FigureCanvasBase` functions
 - Axes3D provides callbacks for the touch functions.

I tested on Python 2.7 and 3.5, with PyQt4 and PySide (both on python 2.7) and PyQt5 (python 2.7 and 3.5), on a Dell XPS 13 running Ubuntu 16.04.

I mentioned that I was working on this in #7789.

### CODE REVIEW

**Comment by GregDMeyer** [comment]
just realized I have some misses on the pep8 compliance--will fix and amend my commit.

**Comment by GregDMeyer** [comment]
fixed the pep8 issues and it should be good to go! (I think the appveyor tests failed due to some issue on their side... it seems there was a problem at runtime?)

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -1137,6 +1140,78 @@ def format_coord(self, xd, yd):
...
+        if self.M is None:
+            return
+
+        w = self._pseudo_w
+        h = self._pseudo_h
+
+        orig_IDs = {t.ID for t in self._touches}
+        e_IDs = {t.ID for t in event.touches}
+        if (len(event.touches) != len(self._touches) or
+                not orig_IDs == e_IDs):
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: `not orig_IDs == eIDs` -> `orig_IDs != eIDs`

**Comment by GregDMeyer** [comment]
Thanks for the comments!

Re: testing, yes, I thought it might be an obstacle that people don't have an easy way to test it. a couple ideas, if it isn't too much of a hassle to set up:

1. VNC from a smartphone or tablet can support passing multitouch events, I believe. (this maybe would be a neat application for this functionality, too, so that if you own a tablet you could hand someone a plot to interact with. I don't have such a device so I don't know how well that would work).
2. There seem to be some ways to simulate a touchscreen, e.g. [this](https://github.com/vi/virtual_touchscreen).

Will go through your comments when I have a chance. cheers

**Comment by thomasaarholt** [comment]
As a heads-up, over at hyperspy/hyperspy-demos#44, we've just realised that using Binder (or public jupyter servers) we can now access jupyter notebooks on our phones. This provides another avenue for testing and especially using multitouch!

I'd love to see this implemented and expanded on!

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -1578,6 +1578,89 @@ def __str__(self):
                                             self.dblclick, self.inaxes)
 
 
+class Touch(LocationEvent):
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: TouchEvent, for consistency with other classes.

File: lib/matplotlib/backend_bases.py [hunk 3]
```diff
@@ -1578,6 +1578,89 @@ def __str__(self):
...
+
+    def __init__(self, name, canvas, x, y, ID, guiEvent=None):
+        """
+        x, y in figure coords, 0,0 = bottom, left
+        """
+        LocationEvent.__init__(self, name, canvas, x, y, guiEvent=guiEvent)
+        self.ID = ID
+
+    def __str__(self):
+        return ("MPL Touch: xy=(%d,%d) xydata=(%s,%s) inaxes=%s ID=%d" %
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: TouchEvent

**Comment by jklymak** [comment]
@GregDMeyer were you planning to come back to this?  The rebase has gotten large, but maybe its straight forward.  Is there a reason you didn't follow up because this was apparently significant work.  

In general this would probably be good to get in.  I would bet 50/50 that macs will have touchscreens in the next few months.  And some PCs have touch.

**Comment by GregDMeyer** [comment]
Unfortunately what happened was that I started grad school and lost track of most of my side projects ;) But I certainly think this would still be a valuable feature, and I'm happy to take another look in the next couple weeks to fix the pull request. I do have a feeling that fixing it up probably won't be that much work. (Also, I have a lot more experience with Python now, so I will definitely want to clean a few things up from my original code).

**Comment by story645** [comment]
My laptop is multi-touch & I'd be happy for the feature/to test it.

**Comment by jklymak** [comment]
@story645 is it OK to assign you to keep an eye on this?  If not, please feel free to unassigned yourself!

**Comment by timhoffm** [comment]
I can test touch screen support as well.

**Comment by jklymak** [comment]
Closing as abandoned.  Thanks for the PR, and sorry it didn't get in.  If you do come back to this, please feel free to request this be re-opened, or open a fresh PR!

**Pull request closed without merging**